src/Authoring/WinRT.SourceGenerator/Generator.cs

```diff
@@ -1,258 +1,342 @@
+        {
+            string outputFile = context.GetWinmdOutputFile();
+            Logger.Log("Writing " + outputFile);
+            var managedPeBuilder = new ManagedPEBuilder(
+                new PEHeaderBuilder(
+                    machine: Machine.I386,
+                    imageCharacteristics: Characteristics.ExecutableImage | Characteristics.Dll | Characteristics.Bit32Machine),
+                new MetadataRootBuilder(metadataBuilder, "WindowsRuntime 1.4"),
+                new BlobBuilder(),
+                flags: CorFlags.ILOnly);
+
+            var peBlob = new BlobBuilder();
+            managedPeBuilder.Serialize(peBlob);
+
+            using var fs = new FileStream(outputFile, FileMode.Create, FileAccess.Write);
+            peBlob.WriteContentTo(fs);
+        }
+
+        private bool CatchWinRTDiagnostics()
+        {
+            string assemblyName = context.GetAssemblyName();
+            WinRTComponentScanner winrtScanner = new(context, assemblyName);
+            winrtScanner.FindDiagnostics();
+            return winrtScanner.Found();
+        }
+
+        public void Generate()
+        {
+            if (CatchWinRTDiagnostics())
+            {
+                Logger.Log("Exiting early -- found errors in authored runtime component.");
+                Logger.Close();
+                Environment.ExitCode = -1;
+                return;
+            }
+
+            try
+            {
+                string assembly = context.GetAssemblyName();
+                string version = context.GetAssemblyVersion();
+                MetadataBuilder metadataBuilder = new MetadataBuilder();
+
+                var writer = new WinRTTypeWriter(
+                    assembly,
+                    version,
+                    metadataBuilder,
+                    Logger);
+
+                WinRTSyntaxReceiver syntaxReceiver = (WinRTSyntaxReceiver)context.SyntaxReceiver;
+                Logger.Log("Found " + syntaxReceiver.Declarations.Count + " types");
+                foreach (var declaration in syntaxReceiver.Declarations)
+                {
+                    writer.Model = context.Compilation.GetSemanticModel(declaration.SyntaxTree);
+                    writer.Visit(declaration);
+                }
+                writer.FinalizeGeneration();
+
+                GenerateWinMD(metadataBuilder);
+                if (!context.ShouldGenerateWinMDOnly())
+                {
+                    GenerateSources();
+                    writer.GenerateWinRTExposedClassAttributes(context);
+                }
+            }
+            catch (Exception e)
+            {
+                Logger.Log(e.ToString());
+                if (e.InnerException != null)
+                {
+                    Logger.Log(e.InnerException.ToString());
+                }
+                Logger.Close();
+                Environment.ExitCode = -2;
+                throw;
+            }
+
+            Logger.Log("Done");

```

Comment:
Note: I normalized whitespaces in this file, **review with whitespace diff off**.
The actual changes are just a few lines of code being added.